klippy/extras/bulk_adc_sensor.py

```diff
@@ -0,0 +1,54 @@
+# Bulk ADC Sensor Support
+#
+# Copyright (C) 2024 Gareth Farrington <gareth@waves.ky>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+# Generic interface for Bulk ADC Sensors that can supply high frequency
+# data from the MCU.
+class BulkSensorAdc(object):
+    # return the MCU the sensor is attached to
+    def get_mcu(self):
+        return None
+    # get the number of samples per second that the sensor is configured for
+    def get_samples_per_second(self):
+        return 1
+    # returns a tuple of the minimum and maximum value of the sensor, used to
+    # detect if a sample value is saturated
+    def get_range(self):
+        return -1, 1
+    # subscribe to data being published from the sensor
+    # callback is called with a list containing (time, counts) tuples:
+    # [(time, counts), (time, counts), ...]
+    # The callback must return True to continue receiving data and False to stop
+    def add_client(self, callback):
+        pass
+
+# Helper for ClockSyncRegression that handles generating timestamps
+# while processing a batch of samples
+class TimestampHelper:
```

Comment:
I agree it would be good to use shared code for the intricate timestamp manipulation that the bulk sensors implement.  However, I think this should be done for all bulk sensors and implemented in bulk_sensor.py .  I'll put together a separate PR for discussion.